compiler/rustc_interface/src/util.rs

```diff
@@ -178,22 +182,26 @@ pub(crate) fn run_in_thread_pool_with_globals<F: FnOnce(CurrentGcx) -> R + Send,
             edition,
             sm_inputs,
             extra_symbols,
-            |current_gcx| {
+            |current_gcx, jobserver_proxy| {
                 // Register the thread for use with the `WorkerLocal` type.
                 registry.register();
 
-                f(current_gcx)
+                f(current_gcx, jobserver_proxy)
             },
         );
     }
 
     let current_gcx = FromDyn::from(CurrentGcx::new());
     let current_gcx2 = current_gcx.clone();
 
+    let proxy = Proxy::new();
```

Comment:
Hang on, won't this somewhat misbehave if you run multiple compiler session inside the same process? They would each think they have an implicit token. There is a reason `GLOBAL_CLIENT_CHECKED` is a static, not a session field.

Reply:
If you run them in parallel yes, but you should already be acquiring an extra token per rustc instance if you're using a jobserver in that case.

Reply:
So you would need to use a separate proxy instance for acquiring the extra token per rustc instance?

Reply:
No you'd use `Client` to get a token before creating the extra `rustc` instance. The proxy is for internal rustc use.

Reply:
If you do it like the following you would deadlock with a jobserver that has 1 token available total:

```rust
for req in requests_stream {
    let token = acquire_token();
    spawn_rustc_instance(token);
}
```

You did have to do something like

```rust
let mut first_request = true;
for req in requests_stream {
    let token = if first_request {
        first_request = false;
        None
    } else {
        Some(acquire_token())
    };
    spawn_rustc_instance(token);
}
```

which is harder than making `Proxy` a static.

Reply:
We could allow a `Proxy` to be passed in to `rustc_interface` to make that easier. Keep in mind `Proxy` cannot be a global as we need its destructor to run.